### DISCUSSION TITLE
Temporarily disable async key event polling

### DISCUSSION THREAD

**Discussion by fritzrehde** [ratatui/ratatui#606] [Discussion]
My question is more related to crossterm, I think, so I filed an issue asking for help there already, see https://github.com/crossterm-rs/crossterm/issues/837. I will repost my message here:

**Is your feature request related to a problem? Please describe.**
In the TUI I am building, I would like to temporarily disable the entire TUI I am painting, execute a command in a subshell that paints its own TUI (e.g. text editor like vim), and then resume my TUI. I use async Rust with Tokio for the key event polling. I have key event polling setup on a separate thread. If I just "leave this polling on" (meaning disable the painting of my TUI, but keep listening to crossterm key events), then my experience in the other TUI (the text editor) is affected by crossterm, as key presses become slugish and take longer to recognize (I'm guessing this is maybe because crossterm is somehow consuming and re-emitting the key presses?). This behaviour is fine though, I shouldn't be listening for key events in my app if I expect the child TUI I am executing to listen to the same key presses. Therefore, I wanted to temporarily (as long as the child TUI is executing) disable key event polling. This is how I currently have that implemented:
```rust
/// A command sent to a polling thread.
enum PollingCommand {
    /// Continue listening/polling for terminal events.
    Listen,
    /// Pause listening/polling for terminal events. Notifies event's sender
    /// once polling has actually been paused.
    Pause(Sender<()>),
}

/// Continuously listens for terminal-related events, and sends relevant events
/// back to the main thread.
async fn poll_terminal_events(event_tx: Sender<Event>, mut polling_rx: Receiver<PollingCommand>) {
    let mut terminal_event_reader = EventStream::new();

    'polling_loop: loop {
        tokio::select! {
            // Wait for receival of a polling command from main event loop thread.
            polling = polling_rx.recv() => match polling {
                Some(PollingCommand::Pause(polling_paused_tx)) => {
                    log::info!("Terminal event listener has been paused.");

                    // Notify sender thread that polling has been paused.
                    let _ = polling_paused_tx.send(()).await;

                    // Wait until another Listen command is received.
                    'wait_for_listen: while let Some(polling) = polling_rx.recv().await {
                        if let PollingCommand::Listen = polling {
                            break 'wait_for_listen;
                        }
                    }

                    log::info!("Terminal event listener is listening again.");
                },
                // Currently already listening for terminal events.
                Some(PollingCommand::Listen) => {},
                // Channel has been closed.
                None => break 'polling_loop,
            },
            // Wait for a terminal event.
            event = terminal_event_reader.next().fuse() => match event {
                Some(Ok(CrosstermEvent::Key(key_event))) => {
                    if event_tx.send(Event::KeyPressed(key_event)).await.is_err() {
                        break 'polling_loop;
                    };
                }
                _ => {}
            }
        }
    }

    log::info!("Shutting down terminal event listener task");
}
```
This does mostly solve the issue I had before. Key events are handled smoothly in the child TUI (text editor). However, some key presses are still exclusively handled by crossterm, i.e. I press a key inside the child TUI and it doesn't register there, but seems to be somehow "saved" by crossterm and then emitted once I leave the child TUI. I am assuming this has something to do with `terminal_event_reader.next().fuse()`. I don't know anything about how `EventStream` works. Is it buffering some events? I was thinking maybe I should clear/delete the entire buffer somehow once I re-enter the `Listen` state, but I don't know how to do that.

**Describe the solution you'd like**
Any help? Maybe I am approaching this problem in a wrong way (with all the channels etc.) and there is a simpler solution. 
I would appreciate any help/comments you have!

**Additional context**
You can check out the exact code in my project: [watchbind](https://github.com/fritzrehde/watchbind/blob/main/src/ui/mod.rs).

**Comment by joshka** [comment]
I have what might be a stupid question (or perhaps just one asked simply and naively enough to appear so).... If you don't want to accept events from crossterm when things should be paused, have you considered not calling `terminal_event_reader.next()` when things are paused?

Or put another way, your loop with the `select!` call returns whenever one of the sides completes, and you're doing nothing to stop the crossterm event stream from trying to read from the terminal. It might be better when you receive a Pause to go into a different branch of the loop that only listens for the Listen event and not for terminal events.

E.g. (naive psuedo code)
```rust
let mut listening = true;
loop {
  if listening {
    select! {
      polling = // ... code to handle pause
      event = // ... code to handle crossterm events
    }
  } else {
    // ... code to handle waiting for listen event
  }
}
```

I don't think there's really anything Ratatui specific about this as Ratatui doesn't have any code that interacts with events in crossterm in the library (only in the example apps), except that you might get a different audience involved that might have done something similar.

**Reply by fritzrehde** [reply]
Very good point, thanks fo r the input. I actually figured that out a couple minutes ago myself, here's my updated code:
```rust
/// Continuously listens for terminal-related events, and sends relevant events
/// back to the main thread.
async fn poll_terminal_events(
    keybindings: Arc<Keybindings>,
    event_tx: Sender<Event>,
    mut polling_rx: Receiver<PollingCommand>,
) {
    let mut terminal_event_reader = EventStream::new();

    'main_loop: loop {
        // Poll terminal events until instructed to pause.
        let polling_paused_tx = 'polling_loop: loop {
            tokio::select! {
                // Wait for receival of a polling command from main event loop thread.
                polling = polling_rx.recv() => match polling {
                    // Once paused, we need to exit the select! statement,
                    // so we don't make further progress on the other
                    // (terminal event polling) future.
                    Some(PollingCommand::Pause(polling_paused_tx)) => break 'polling_loop polling_paused_tx,
                    // Currently already listening for terminal events.
                    Some(PollingCommand::Listen) => {},
                    // Channel has been closed.
                    None => break 'main_loop,
                },
                // Wait for a terminal event.
                event = terminal_event_reader.next().fuse() => match event {
                    Some(Ok(CrosstermEvent::Key(key_event))) => {
                        if event_tx.send(Event::KeyPressed(key_event)).await.is_err() {
                            break 'main_loop;
                        };
                    }
                    _ => {}
                }
            }
        };

        log::info!("Terminal event listener has been paused.");

        // Notify sender thread that polling has been paused.
        let _ = polling_paused_tx.send(PollingPaused).await;

        // Wait until another Listen command is received.
        'wait_for_listen: while let Some(polling) = polling_rx.recv().await {
            if let PollingCommand::Listen = polling {
                break 'wait_for_listen;
            }
        }

        log::info!("Terminal event listener is listening again.");
    }

    log::info!("Shutting down terminal event listener task");
}
```
Rust allows returning values from loops, which makes this quite elegant. However, unfortunately, this changes nothing about the behaviour I have observed. This is what you meant though, @joshka, right?

**Comment by fritzrehde** [comment]
Yes, I do know that this issue is not related to ratatui directly, but I thought there might some people here that have either encountered this issue as well, or know of a project that implements a similar feature where I could have a look.

**Comment by fritzrehde** [comment]
@sxyazi I thought I might tag you, as you seem to be the creator and maintainer of [`yazi`](https://github.com/sxyazi/yazi). I just found out that `yazi` also uses crossterm, ratatui and async rust with Tokio, so maybe you can guide me in the right direction, especially since `yazi` also supports opening files. Would appreciate any help!

**Reply by fritzrehde** [reply]
I found https://github.com/sxyazi/yazi/blob/73941c22917469bdd7877877f80981438af653fd/yazi-fm/src/signals.rs#L79, which looks quite similar to my code.

**Comment by joshka** [comment]
Taking a superficial read of the crossterm code, I suspect the answer might be to drop the EventStream. Otherwise, there's a thread that continuously polls for a new event and reads the event when one is available
https://github.com/crossterm-rs/crossterm/blob/master/src/event/stream.rs#L44-L59

Just a guess at this point - the internals of crossterm events are pretty complex.

**Reply by fritzrehde** [reply]
Cool, dropping it before going into the "paused" state actually seems to have fixed it (at least I can't manually reproduce the error anymore). The weird thing is that I already tried this yesterday, and it didn't work. Maybe the other fix you suggested was required as well for it to work. Thank you so much for your time and help!

**Reply by joshka** [reply]
Sweet :) Glad that I could help.

**Comment by sxyazi** [comment]
Sorry I'm late, I'm glad to see that your question has been resolved!

**Reply by fritzrehde** [reply]
No worries! Just a quick question on your use of the `EventStream`. Do you call your `spawn_crossterm_task` again everytime you start listening for crossterm events? That way you would ensure a new `EventStream` is created everytime.

**Reply by sxyazi** [reply]
Yeah, I will discard the `EventStream` temporarily when it needs to be disabled and create a new one when it needs to be restored.

**Reply by fritzrehde** [reply]
Cool, that's what I thought, thanks!

**Comment by fritzrehde** [comment]
@sxyazi I actually would have another quick question that's very related to my question here: So I seem to have figured out the crossterm key handling now, but what `yazi` still does better than `watchbind` (my project) IMO is the transition from the host TUI (yazi's or watchbind's) to the child TUI (e.g. the text editor). For watchbind, I basically completely unpaint the TUI, and only then execute the child TUI. This means that for a split second, the command line, from which watchbind was called, is shown. But for yazi this doesn't seem to be the case. Any hints on how you did that? Did you just not completely hide yazi's TUI, and just let the child TUI "override" yazi's TUI? Thanks!

**Reply by sxyazi** [reply]
When the need to display the child UI, I [destroy Term](https://github.com/sxyazi/yazi/blob/73941c22917469bdd7877877f80981438af653fd/yazi-fm/src/app.rs#L101) immediately, without waiting for any events, and I stop the `EventStream` at the same time as destroying, then a child process is spawned.

**Reply by fritzrehde** [reply]
Does destroying the terminal mean preventing further UI updates to the terminal user interface, or "completely exiting the TUI" like I do here: https://github.com/fritzrehde/watchbind/blob/fab908d7e9a5d8f72b371a318102a88607ce370b/src/ui/terminal_manager.rs#L43C8-L43C8

**Reply by sxyazi** [reply]
Both. When `term` is set to `None`, two things are done:

1. Destroy the TUI, https://github.com/sxyazi/yazi/blob/73941c22917469bdd7877877f80981438af653fd/yazi-shared/src/term/term.rs#L72
2. Ignore all UI updates, https://github.com/sxyazi/yazi/blob/73941c22917469bdd7877877f80981438af653fd/yazi-fm/src/app.rs#L77

**Reply by joshka** [reply]
I've mentioned this elsewhere that I really like the `Term` abstraction that is used in yazi, particularly the start() and `Drop` implementation. It's worth having a read of how that works as a whole. https://github.com/sxyazi/yazi/blob/main/yazi-shared/src/term/term.rs

**Reply by fritzrehde** [reply]
Just to confirm that I understood correctly: The feature I described that I liked in yazi (seamless transition from own TUI to child TUI) is, firstly, on purpose (and not just a coincidence) and, secondly, due to the `Term` implementation?

**Reply by sxyazi** [reply]
Yes I think that's the way to quickly respond to user requests.

It makes me curious why there is a delay in watchbind when transitioning, is it waiting for other events (such as the next rendering)?

**Reply by fritzrehde** [reply]
Actually, I just screen recorded while opening files in `yazi` as well (I did so for `watchbind` earlier), and when looking at the recording frame by frame you can indeed see that `yazi` does briefly, for a couple of frames, return to the user's terminal (from which yazi was called), and only then paints the editor's TUI. The same goes for transitioning from the editor back to yazi, there is a small time spent on the user's terminal before yazi's TUI gets painted. I checked out some other popular terminal file managers out of curiousity (`lf` and `ranger`), and they both have the same characteristics.
Of course, we are talking about a couple frames mostly invisible to the eye, but I am, nevertheless, curious if there are solutions to mitigate this.
I actually fixed this problem partially in my app watchbind, I think. When transitioning from watchbind's TUI to the child TUI (editor), there is no longer this short return to the user's terminal. This is because I don't execute crossterm's `LeaveAlternateScreen` when hiding watchbind's TUI. By not leaving this alternate screen, the child TUI seems to just paint over watchbind's TUI, so there is no return to the user's terminal. You can check out my PR fix (https://github.com/fritzrehde/watchbind/pull/71) and the accompanying issue (https://github.com/fritzrehde/watchbind/issues/67).
However, I didn't manage to fix it for transitioning from the child TUI back to watchbind's TUI. I think this isn't really possible, because it's not like we can have a "layer" underneath the child TUI, which then gets shown again once the child TUI exits. Watchbind's TUI gets overriden by the child, and we need to repaint all of watchbind's TUI after the child TUI exits. But please let me know if you have any ideas on this. We could discuss this at https://github.com/fritzrehde/watchbind/issues/72, here, or in a new ratatui issue/discussion, up to you.

**Reply by kdheepak** [reply]
I'll have to read through your source code to see exactly what you are doing, but I don't think there's a way to work around the fact that a child TUI calls `LeaveAlternateScreen`. 

What is wrong with repainting after the child TUI returns control?

Here's a screen recording through vhs of me doing something similar in taskwarrior-tui:

https://user-images.githubusercontent.com/1813121/280558722-dbb66113-570c-483d-82df-c6adf74a5a02.mp4

Here's the commands I ran:

```
Set Shell zsh
Set FontSize 12
Set Width 900
Set Height 600
Set Theme "Catppuccin Mocha"
Output out.mp4
Sleep 2s
Type "tt"
Enter
Sleep 1.5s
Type "e"
Sleep 5s
Type "jjkk"
Sleep 5s
Type ":q"
Enter
Sleep 5s
Type "q"
Sleep 5s
```

I use `LeaveAlternateScreen`, and execute a command that starts a subprocess to start vim, and once vim finishes and returns control back to my program I run `EnterAlternateScreen` again. 

The child TUI in this case is vim, which calls its own `EnterAlternateScreen` + `LeaveAlternateScreen` but that doesn't have to be the case. Other child TUIs can run in the current screen without ever entering the alternate screen. 

The delay like you mentioned is probably just a frame or two and not really visible to me in normal use. I also start the event handler before calling `EnterAlternateScreen` on returning from the child TUI so if a user presses any keys in that time frame they are buffered and for the user it is mostly a seamless experience.

**Reply by fritzrehde** [reply]
Yes, you're completely right, the delay is just a frame or two, and is not really visible in normal use to me either. I did download your screen recording and went throug it frame by frame, and you can see it exit taskwarrior and return to the calling terminal before painting the editor. For your exact screen recording, the transition is not noticeable at all imo, however, I think that's partially because your screen is mostly blank (besides the `tt`) anyways. If it were filled completely with colorful text, I tend to notice it some more, in my experience. I acknowledge that I'm being super nit-picky here though, so just ignoring it is totally fine :) 
If you want to check out my implementation, just take a look at https://github.com/fritzrehde/watchbind/blob/main/src/ui/terminal_manager.rs.

**Reply by fritzrehde** [reply]
What exactly do you mean with:
```
Other child TUIs can run in the current screen without ever entering the alternate screen.
```
Do you have an example of such a TUI command? My understanding of a TUI is that it always paints its own, entire screen. Or do you mean something like fzf in non-full such as `fzf --height 10`. I have indeed not taken that into account. That's actually a great point, and I might have to think about whether I want to support that or not. Any other examples with such behaviour?

**Reply by kdheepak** [reply]
I was thinking of fzf. But I’m sure there’s more. I get the sense there are some apps in the charm organization that are interactive without starting an alternative screen.

**Comment by kdheepak** [comment]
The [ratatui-async-template](https://github.com/ratatui-org/ratatui-async-template) has a [`tui.stop()`](https://github.com/ratatui-org/ratatui-async-template/blob/c792a2a5d5cca72a47844acec99ef7753af007f6/template/src/tui.rs#L154-L169) method, which  uses `tokio_util::sync::CancellationToken` to [`break` out of the `while` loop](https://github.com/ratatui-org/ratatui-async-template/blob/c792a2a5d5cca72a47844acec99ef7753af007f6/template/src/tui.rs#L108-L110). 

I've used this successfully to `tui.stop()` the event loop, start vim to edit some data, and to `tui.start()` the event loop again after. [`tui.start()`](https://github.com/ratatui-org/ratatui-async-template/blob/c792a2a5d5cca72a47844acec99ef7753af007f6/template/src/tui.rs#L91) creates a new `while` loop as an async task with a new `CancellationToken`.
